homeassistant/components/glinet/config_flow.py

```diff
@@ -0,0 +1,171 @@
+
+import logging
+from typing import Any
+
+from gli_py import GLinet
+import voluptuous as vol
+
+from homeassistant import config_entries
+from homeassistant.components.device_tracker.const import (
+    CONF_CONSIDER_HOME,
+    DEFAULT_CONSIDER_HOME,
+)
+from homeassistant.const import CONF_API_TOKEN, CONF_HOST, CONF_PASSWORD
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.data_entry_flow import FlowResult
+from homeassistant.exceptions import HomeAssistantError
+from homeassistant.helpers.device_registry import format_mac
+
+from .const import DOMAIN
+
+
+_LOGGER = logging.getLogger(__name__)
+
+STEP_USER_DATA_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_HOST, default="http://192.168.8.1"): str,
+        vol.Required(CONF_PASSWORD, default="goodlife"): str,
+    }
+)
+
+
+class TestingHub:
+    """Testing class to test connection and authentication."""
+
+    def __init__(self, host: str) -> None:
+        """Initialize."""
+        self.host: str = host
+        self.router: GLinet = GLinet(base_url=self.host + "/cgi-bin/api/", sync=False)
+        self.router_model: str = ""
+        self.router_mac: str = ""
+
+    async def connect(self) -> bool:
+        """Test if we can communicate with the host."""
+        try:
+            res = await self.router.router_model()
+            self.router_model = res["model"]
+            return True
+        except ConnectionError:
+            _LOGGER.error(
+                "Failed to connect to %s, is it really a GL-inet router?", self.host
+            )
+        except TypeError:
+            _LOGGER.error(
+                "Failed to parse router response to %s, is it the right firmware version?",
+                self.host,
+            )
+        return False
+
+    async def authenticate(self, password: str) -> bool:
+        """Test if we can authenticate with the host."""
+        try:
+            await self.router.async_login(password)
+            res = await self.router.router_mac()
+            self.router_mac = res["factorymac"]
+        except ConnectionRefusedError:
+            _LOGGER.error("Failed to authenticate with Gl-inet router during testing")
+        return self.router.logged_in
+
+
+async def validate_input(hass: HomeAssistant, data: dict[str, Any]) -> dict[str, Any]:
+    """Validate the user input allows us to connect.
+
+    Data has the keys from STEP_USER_DATA_SCHEMA with values provided by the user.
+    """
+
+    hub = TestingHub(data[CONF_HOST])
+
+    if not await hub.connect():
+        raise CannotConnect
+
+    if not await hub.authenticate(data[CONF_PASSWORD]):
+        raise InvalidAuth
+
+    # Return info that you want to store in the config entry.
+    return {
+        "title": "GL-inet " + hub.router_model,
+        "mac": hub.router_mac,
+        "data": {
+            CONF_HOST: data[CONF_HOST],
+            CONF_API_TOKEN: hub.router.token,
+            CONF_PASSWORD: data[CONF_PASSWORD],
```

Comment:
Why store the password, you have the token?

Reply:
I tried without, but unfortunately there was no way to handle token renewal without storing the password.